content/blog/2018-10-09-jstor.md

```diff
@@ -0,0 +1,512 @@
+to learn how to navigate within the document via XPATH. 
+
+Another thing I took up
+during this re-write was regularly benchmarking my functions. Since I lacked 
+the technical expertise to know up front which approach would be faster,
+I benchmarked possibilities to figure it out. The following benchmark 
+compares three versions of extracting the text from the node named `volume`.
+Especially the first two seemed very similar to me and it was not at all 
+obvious, which one would be faster.
+
+```r
+fun1 <- function() xml_find_all(test_file, "//volume/.//text()") %>% as.character()
+fun2 <- function() xml_find_all(test_file, "//volume") %>% xml_text()
+fun3 <- function() {
+  xml_find_all(test_file, "front") %>%
+    xml_child("article-meta") %>%
+    xml_child("volume") %>%
+    xml_text()
+}
+
+microbenchmark::microbenchmark(fun1(), fun2(), fun3(), times = 1000)
+# Unit: microseconds
+# expr     min       lq     mean   median       uq      max neval
+# fun1() 267.684 290.5885 322.8992 301.4325 318.9060 2705.459  1000
+# fun2() 233.723 250.3840 287.1126 263.1725 273.2180 2716.416  1000
+# fun3() 686.542 725.0925 822.2103 763.4240 798.1155 3534.849  1000
+```
+Although the unit is microseconds here, the difference between `fun2` and 
+`fun3` is quite substantial, if you want to do it 500,000 times: 4.5 minutes. If
+you have to extract elements often (which I was doing), then this quickly adds 
+up. What is the reason for this big difference?
+I'd guess that `fun2` is fastest since it only calls two functions, and not 
+four, and because `xml_text()` probably converts the content directly to 
+character in C++, which is faster than calling `as.character` afterwards. Still, 
+I am not a computer scientist, so I am not entirely sure about that.
+
+
+After rewriting and expanding the functionality, I was still not happy, however,
+since the code was slower than I thought it should be. At this point,
+I dug deeper again, using the package
+[profvis](https://rstudio.github.io/profvis/) to find bottlenecks. Oddly
+enough, I had introduced a bottleneck right at the beginning: I was using
+`data_frame` (which is equivalent to `tibble()`) to create the object my 
+function would return. Unfortunately, `tibble` (and `data.frame` as well) are
+quite complex functions. They do type checking and other things, and if you do
+this repeatedly, it quickly adds up. 
+
+You can see the difference yourself:
+
+```r
+microbenchmark::microbenchmark(
+  tibble = tibble::tibble(a = rnorm(10), b = letters[1:10]),
+  new_tibble = tibble::new_tibble(list(a = rnorm(10), b = letters[1:10]))
+)
+```
+
+```
+## Unit: microseconds
+##        expr     min       lq     mean   median      uq      max neval
+##      tibble 332.902 479.7430 624.1354 621.7795 716.437 1944.658   100
+##  new_tibble  76.833 106.5485 155.4881 151.2890 172.570  974.428   100
+```
+
+The difference is not huge (~50sec for 500,000 documents), but my actual layout
```

Comment:
feeling naive: which columns in the above show "(~50sec for 500,000 documents)" ? Similar for the "40 minutes more for 500,000 documents" below